src/app/clusters/occupancy-sensor-server/OccupancySensingCluster.cpp

```diff
@@ -0,0 +1,333 @@
+
+        // If the elapsed time is already greater than or equal to the new hold time,
+        // the timer should have already fired. Cancel the current timer and immediately
+        // set the state to unoccupied.
+        if (elapsedTime >= newHoldTimeSeconds)
+        {
+            mTimerDelegate->CancelTimer(this);
+            DoSetOccupancy(false);
+        }
+        else
+        {
+            // Otherwise, restart the timer for the remaining duration.
+            System::Clock::Timeout remainingTime = newHoldTimeSeconds - elapsedTime;
+            mTimerDelegate->StartTimer(this, remainingTime);
+        }
+    }
+
+    if (mDelegate)
+    {
+        mDelegate->OnHoldTimeChanged(holdTime);
+    }
+
+    if (mContext != nullptr)
+    {
+        // There's not much we can do if persistence fails here, so we ignore the return value.
+        // The application must ensure that persistence is working correctly.
+        RETURN_SAFELY_IGNORED mContext->attributeStorage.WriteValue(
+            { mPath.mEndpointId, OccupancySensing::Id, Attributes::HoldTime::Id },
+            { reinterpret_cast<const uint8_t *>(&mHoldTime), sizeof(mHoldTime) });
+    }
+
+    return Protocols::InteractionModel::Status::Success;
+}
+
+void OccupancySensingCluster::SetOccupancy(bool occupied)
+{
+    if (mHasHoldTime && mTimerDelegate)
+    {
+        if (occupied)
+        {
+            mTimerDelegate->CancelTimer(this);
+            DoSetOccupancy(true);
+        }
+        else
+        {
+            mTimerStartedTimestamp = mTimerDelegate->GetCurrentMonotonicTimestamp();
+            mTimerDelegate->StartTimer(this, System::Clock::Seconds16(mHoldTime));
+        }
+        return;
+    }
+
+    DoSetOccupancy(occupied);
+}
+
+void OccupancySensingCluster::TimerFired()
+{
+    DoSetOccupancy(false);
+}
+
+void OccupancySensingCluster::DoSetOccupancy(bool occupied)
+{
+    BitMask<OccupancySensing::OccupancyBitmap> newOccupancy =
+        occupied ? OccupancySensing::OccupancyBitmap::kOccupied : static_cast<OccupancySensing::OccupancyBitmap>(0);
+
+    if (mOccupancy == newOccupancy)
+    {
+        return;
+    }
+
+    mOccupancy = newOccupancy;
+    NotifyAttributeChanged(Attributes::Occupancy::Id);
+
+    if (mDelegate)
+    {
+        mDelegate->OnOccupancyChanged(occupied);
+    }
+
+    if (mContext != nullptr)
+    {
+        Events::OccupancyChanged::Type event;
+        event.occupancy = mOccupancy;
+        mContext->interactionContext.eventsGenerator.GenerateEvent(event, mPath.mEndpointId);
+    }
+}
+
+bool OccupancySensingCluster::IsOccupied() const
+{
+    return mOccupancy.Has(OccupancySensing::OccupancyBitmap::kOccupied);
+}
+
+uint16_t OccupancySensingCluster::GetHoldTime() const
+{
+    return mHoldTime;
+}
+
+const OccupancySensing::Structs::HoldTimeLimitsStruct::Type & OccupancySensingCluster::GetHoldTimeLimits() const
+{
+    return mHoldTimeLimits;
```

Comment:
Same as above, this Struct should probably be optional, and only be populated on the presence of the HoldTimeAttribte in the cluster's OptionalAttributesSet